src/libmongoc/tests/unified/entity-map.c

```diff
@@ -0,0 +1,748 @@
+   bool ret = false;
+
+   bson_iter_init (&iter, bson);
+   if (!bson_iter_next (&iter)) {
+      test_set_error (error, "Empty entity");
+      goto done;
+   }
+
+   entity_type = bson_iter_key (&iter);
+   bson_iter_bson (&iter, &entity_bson);
+   if (bson_iter_next (&iter)) {
+      test_set_error (error,
+                      "Extra field in entity: %s: %s",
+                      bson_iter_key (&iter),
+                      tmp_json (bson));
+      goto done;
+   }
+
+   if (0 == strcmp (entity_type, "client")) {
+      entity = entity_client_new (&entity_bson, error);
+   } else if (0 == strcmp (entity_type, "database")) {
+      entity = entity_database_new (entity_map, &entity_bson, error);
+   } else if (0 == strcmp (entity_type, "collection")) {
+      entity = entity_collection_new (entity_map, &entity_bson, error);
+   } else if (0 == strcmp (entity_type, "session")) {
+      entity = entity_session_new (entity_map, &entity_bson, error);
+   } else {
+      test_set_error (
+         error, "Unknown entity type: %s: %s", entity_type, tmp_json (bson));
+      goto done;
+   }
+
+   if (!entity) {
+      goto done;
+   }
+
+   LL_FOREACH (entity_map->entities, entity_iter)
+   {
+      if (0 == strcmp (entity_iter->id, entity->id)) {
+         test_set_error (
+            error, "Attempting to create duplicate entity: '%s'", entity->id);
+         entity_destroy (entity);
+         goto done;
+      }
+   }
+
+   ret = true;
+done:
+   if (!ret) {
+      entity_destroy (entity);
+   } else {
+      LL_PREPEND (entity_map->entities, entity);
+   }
+   return ret;
+}
+
+static void
+entity_destroy (entity_t *entity)
+{
+   event_t *event;
+   event_t *tmp;
+
+   if (!entity) {
+      return;
+   }
+
+   BSON_ASSERT (entity->type);
+
+   if (0 == strcmp ("client", entity->type)) {
+      mongoc_client_t *client;
+
+      client = (mongoc_client_t *) entity->value;
+      mongoc_client_destroy (client);
+   } else if (0 == strcmp ("database", entity->type)) {
+      mongoc_database_t *db;
+
+      db = (mongoc_database_t *) entity->value;
+      mongoc_database_destroy (db);
+   } else if (0 == strcmp ("collection", entity->type)) {
+      mongoc_collection_t *coll;
+
+      coll = (mongoc_collection_t *) entity->value;
+      mongoc_collection_destroy (coll);
+   } else if (0 == strcmp ("session", entity->type)) {
+      mongoc_client_session_t *sess;
+
+      sess = (mongoc_client_session_t *) entity->value;
+      mongoc_client_session_destroy (sess);
+   }
+
+   LL_FOREACH_SAFE (entity->events, event, tmp)
+   {
+      event_destroy (event);
+   }
+
+   bson_destroy (entity->ignore_command_monitoring_events);
+   bson_free (entity->type);
+   bson_free (entity->id);
+   bson_free (entity);
+}
+
+entity_t *
+entity_map_get (entity_map_t *entity_map, const char *id, bson_error_t *error)
+{
+   entity_t *entity;
+   LL_FOREACH (entity_map->entities, entity)
+   {
+      if (0 == strcmp (entity->id, id)) {
+         break;
```

Comment:
nit, I have no strong feelings about it either way but if the entity was returned here then the NULL check wouldn't be needed below. Feel free to ignore this though.

Reply:
That seems like a good simplification. Done.